### DISCUSSION TITLE
Enable fetch alerts for Prometheus compatible datasources

### DISCUSSION THREAD

**Discussion by dmitryk-dk** [grafana/grafana#68169] [Plugins]
We are developing our data sources and have a problem loading alerting rules from the data source.
If you click on alerting -> alerting rules request from our data source implementation goes to `api/prometheus/grafana/api/v1/rules` API. This API returns only created alerts in the grafana and does not retrieve information from datasource itself. 
But if you choose prometheus or loki as data source, it sends the requests to the
`api/prometheus/{datasource_id}/api/v1/rules` API.

From the code side, there is a constant defined that prevents make requests to different datasource:

From the code side, there is a constant defined that prevents make requests to different datasource: 

```typescript
export const RulesDataSourceTypes: string[] = [DataSourceType.Loki, DataSourceType.Prometheus];
```

The chain of function calls goes like this:
1. `fetchAllPromAndRulerRulesAction` https://github.com/grafana/grafana/blob/64ee42d01ef779647e04fc88aa46673db38773f3/public/app/features/alerting/unified/state/actions.ts#L362 
gets data sources from the call of the `getAllRulesSourceNames`
https://github.com/grafana/grafana/blob/64ee42d01ef779647e04fc88aa46673db38773f3/public/app/features/alerting/unified/state/actions.ts#L370

2. `getAllRulesSourceNames` calls `getRulesDataSources`
https://github.com/grafana/grafana/blob/329d94044891b6a7ec0df5cf9b3d84c0ccbb38fd/public/app/features/alerting/unified/utils/datasource.ts#L28

3. `getRulesDataSource` has filter https://github.com/grafana/grafana/blob/329d94044891b6a7ec0df5cf9b3d84c0ccbb38fd/public/app/features/alerting/unified/utils/datasource.ts#L34 where only `prometheus` or `loki` can handle those requests, because 
```typescript
export const RulesDataSourceTypes: string[] = [DataSourceType.Loki, DataSourceType.Prometheus];
```

Even if we can use data source type `prometheus` in the frontend part from the build info function 
``` typescript
async getBuildInfo() {
    try {
      const buildInfo = await discoverDataSourceFeatures({ url: this.url, name: this.name, type: 'prometheus' });
      return buildInfo;
    } catch (error) {
      // We don't want to break the rest of functionality if build info does not work correctly
      return undefined;
    }
  }
  ```
  
  All request also handles at the backend part of the grafana and returns errors messages because
  1. Only `prometheus` and `loki` defined as map keys 
  https://github.com/grafana/grafana/blob/1400a0a3f7a5b5b3c5b7258a3b5bb037b654f599/pkg/services/ngalert/api/lotex_prom.go#LL18C1-L18C1
  2. The same when it tries to call ruler 
  https://github.com/grafana/grafana/blob/1400a0a3f7a5b5b3c5b7258a3b5bb037b654f599/pkg/services/ngalert/api/lotex_ruler.go#L35
  
So, in any case, we can't call `api/prometheus/grafana/api/v1/rules` API when using our data source.

The proposed solution is to add some type of `prometheus-compatible` data source or when developers defines `plugin.json` that the data source supports alerting enable requests to the `api/prometheus/grafana/api/v1/rules` API

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82677. Feel free to continue the discussion around this there. Thank you!
